src/conventions/h-group/take-action.js

```diff
@@ -189,22 +189,38 @@ export function take_action(game) {
 		find_best_playable(game, playable_cards, playable_priorities) :
 		{ priority: -1, best_playable_card: undefined };
 
-	// Playing into finesse/bluff
-	if (playable_cards.length > 0 && priority === 0)
+	// Bluffs should never be deferred as they can lead to significant desync with human players
+	if (playable_cards.length > 0 && priority === 0 && playable_cards.some(c => common.thoughts[c.order].bluffed))
 		return { tableID, type: ACTION.PLAY, target: best_playable_card.order };
 
+	const has_finesse_plays = playable_cards.length > 0 && priority === 0;
 	// Unlock next player
-	if (urgent_actions[ACTION_PRIORITY.UNLOCK].length > 0)
+	if (!has_finesse_plays && urgent_actions[ACTION_PRIORITY.UNLOCK].length > 0)
 		return urgent_actions[ACTION_PRIORITY.UNLOCK][0];
 
-	// Urgent save for next player
-	for (let i = 1; i < actionPrioritySize; i++) {
-		const action = urgent_actions[i].find(action => state.clue_tokens > 0 || (action.type !== ACTION.RANK && action.type !== ACTION.COLOUR));
-
-		if (action)
-			return action;
+	// TODO: We should be able to delay a finesse for any urgent action. This will require
+	// the other players being able to recognize the urgent action in update-turn.js.
+	if (has_finesse_plays) {
```

Comment:
Is the intention to prioritize all urgent actions to the next player before playing into a finesse? This combination of if statement + loop condition seems unnecessary if so: why not just move the loop in front of playing into a finesse?

Reply:
> Is the intention to prioritize all urgent actions to the next player before playing into a finesse?

Eventually, yes, I'd like to do that. But we need to make sure that everyone can recognize that it was absolutely urgent.

> This combination of if statement + loop condition seems unnecessary if so: why not just move the loop in front of playing into a finesse?

Ah, good point! I got here from starting with a condition in the loop which ended up being implementable by a single line so I pulled it out of the loop, but just pulling it out in front of playing a finesse is much cleaner.